src/controller/CHIPDeviceControllerFactory.cpp

```diff
@@ -148,12 +148,13 @@ CHIP_ERROR DeviceControllerFactory::InitSystemState(FactoryInitParams params)
 #endif
                                                             ));
 
-    stateParams.sessionMgr                                    = chip::Platform::New<SessionManager>();
-    SimpleSessionResumptionStorage * sessionResumptionStorage = chip::Platform::New<SimpleSessionResumptionStorage>();
-    stateParams.sessionResumptionStorage                      = sessionResumptionStorage;
-    stateParams.exchangeMgr                                   = chip::Platform::New<Messaging::ExchangeManager>();
-    stateParams.messageCounterManager                         = chip::Platform::New<secure_channel::MessageCounterManager>();
-    stateParams.groupDataProvider                             = params.groupDataProvider;
+    stateParams.sessionMgr = chip::Platform::New<SessionManager>();
+    auto sessionResumptionStorage =
+        std::unique_ptr<SimpleSessionResumptionStorage, Platform::Deleter<chip::SessionResumptionStorage>>(
```

Comment:
Please use Platform::MakeUnique here.  It would be much shorter and simpler.

Reply:
UniquePtr<SimpleSessionResumptionStorage> isn't convertible to UniquePtr<SessionResumptionStorage>

That's the issue. There are three solutions:

1) Use std::unique_ptr<SimpleSessionResumptionStorage, Platform::Deleter<chip::SessionResumptionStorage>> as done here

2) Do the conversion manually (UniquePtr<chip::SessionResumptionStorage>(sessionResumptionStorage.release());

3) Add a converting constructor for Deleter<> (like std::default_delete has). We could add one, but then **we must be sure to use virtual destructors everywhere it is used**. I wasn't sure about that assumption, as virtual destructors have a cost and we're sensitive to those.

Reply:
So, does `DeviceControllerSystemStateParams::sessionResumptionStorage` really need to be of type `SessionResumptionStorage`? This struct is only used by the factory no?

Reply:
Is the goal to _eventually_ be able to provide through `stateParams` a `SessionResumptionStorage` that can then be injected into the call to `InitSystemState`? That clearly isn't permitted today...

Reply:
It seems better to me to leave open the possibility of actually making use of the polymorphism, by allowing injection.

Reply:
I assumed this was tracked by https://github.com/project-chip/connectedhomeip/issues/17412, so was just happy to see any fix that removes the leak right now.